internal/consensus/state_test.go

```diff
@@ -2197,6 +2201,79 @@ func TestExtendVote(t *testing.T) {
 	mock.AssertExpectationsForObjects(t, m)
 }
 
+// TestCorrectABCIRound checks if ABCI receives correct round number in requests.
+func TestCorrectABCIRound(t *testing.T) {
+	config := configSetup(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	m := abcimocks.NewApplication(t)
+
+	cs1, vss := makeState(ctx, t, makeStateArgs{config: config, application: m})
+	height := cs1.Height
+
+	proTxHashMap := make(map[string]struct{})
+	for _, vs := range vss {
+		pth, _ := vs.GetProTxHash(ctx)
+		proTxHashMap[pth.String()] = struct{}{}
+	}
+
+	proposalCh := subscribe(ctx, t, cs1.eventBus, types.EventQueryCompleteProposal)
+	newRoundCh := subscribe(ctx, t, cs1.eventBus, types.EventQueryNewRound)
+	proTxHash, err := cs1.privValidator.GetProTxHash(ctx)
+	require.NoError(t, err)
+	voteCh := subscribeToVoter(ctx, t, cs1, proTxHash)
+
+	started := false
+	// Doing 4 empty rounds, to ensure round != 0 and cs1 is proposer
+	var round int32
+	for round = 0; round < 4; round++ {
+		t.Logf("Starting round %d", round)
+		// we are proposer, so expect proposal
+		if round%4 == 0 {
+			mockProposerApplicationCalls(t, m, round, false)
+		}
+
+		// start the test
+		if started == false {
```

Comment:
you can simplify this code a bit, if replace the condition on `round == 0`, then `started` variable would not be needed